processor/intervalprocessor/partitioned_processor.go

```diff
@@ -0,0 +1,284 @@
+					cloneHistogram := mClone.Histogram()
+
+					aggregateDataPointsByPartition(histogram.DataPoints(), cloneHistogram.DataPoints(), metricID, p.histogramLookup)
+					return true
+				case pmetric.MetricTypeExponentialHistogram:
+					expHistogram := m.ExponentialHistogram()
+
+					if expHistogram.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					mClone, metricID := p.getOrCloneMetric(rm, sm, m)
+					cloneExpHistogram := mClone.ExponentialHistogram()
+
+					aggregateDataPointsByPartition(expHistogram.DataPoints(), cloneExpHistogram.DataPoints(), metricID, p.expHistogramLookup)
+					return true
+				default:
+					errs = errors.Join(fmt.Errorf("invalid MetricType %d", m.Type()))
+					return false
+				}
+			})
+			return sm.Metrics().Len() == 0
+		})
+		return rm.ScopeMetrics().Len() == 0
+	})
+
+	if err := p.nextConsumer.ConsumeMetrics(ctx, md); err != nil {
+		errs = errors.Join(errs, err)
+	}
+
+	return errs
+}
+
+func aggregateDataPointsByPartition[DPS metrics.DataPointSlice[DP], DP metrics.DataPoint[DP]](dataPoints DPS, mCloneDataPoints DPS, metricID identity.Metric, dpLookup []map[identity.Stream]DP) {
+	for i := 0; i < dataPoints.Len(); i++ {
+		dp := dataPoints.At(i)
+
+		streamID := identity.OfStream(metricID, dp)
+		partition := streamID.Hash().Sum64() % uint64(len(dpLookup))
+		existingDP, ok := dpLookup[partition][streamID]
+		if !ok {
+			dpClone := mCloneDataPoints.AppendEmpty()
+			dp.CopyTo(dpClone)
+			if dpLookup[partition] == nil {
+				dpLookup[partition] = make(map[identity.Stream]DP)
+			}
+			dpLookup[partition][streamID] = dpClone
+			continue
+		}
+
+		// Check if the datapoint is newer
+		if dp.Timestamp() > existingDP.Timestamp() {
+			dp.CopyTo(existingDP)
+			continue
+		}
+
+		// Otherwise, we leave existing as-is
+	}
+}
+
+func (p *PartitionedProcessor) exportMetrics(partition int) {
+	md := func() pmetric.Metrics {
+		p.stateLock.Lock()
+		defer p.stateLock.Unlock()
+
+		// ConsumeMetrics() has prepared our own pmetric.Metrics instance ready for us to use
+		// Take it and clear replace it with a new empty one
+		out := p.md[partition]
+		p.md[partition] = pmetric.NewMetrics()
+
+		// Clear all the lookup references
+		clear(p.rmLookup[partition])
+		clear(p.smLookup[partition])
+		clear(p.mLookup[partition])
+		clear(p.numberLookup[partition])
+		clear(p.histogramLookup[partition])
+		clear(p.expHistogramLookup[partition])
+		clear(p.summaryLookup[partition])
+
+		return out
+	}()
+
+	if err := p.nextConsumer.ConsumeMetrics(p.ctx, md); err != nil {
+		p.logger.Error("Metrics export failed", zap.Error(err))
+	}
+}
+
+func (p *PartitionedProcessor) getOrCloneMetric(rm pmetric.ResourceMetrics, sm pmetric.ScopeMetrics, m pmetric.Metric) (pmetric.Metric, identity.Metric) {
+	// Find the ResourceMetrics
+	resID := identity.OfResource(rm.Resource())
+	partition := resID.Hash().Sum64() % uint64(len(p.rmLookup))
+	rmClone, ok := p.rmLookup[partition][resID]
```

Comment:
unless I'm totally mistaken, you can't do it like that.

iiuc the partition must be determined using `identity.Stream` only, or you end up writing to different partitions here, as `resID.Sum64() % 60` is nearly guaranteed to be different than `streamID.Sum64() % 60`.

in `exportMetrics` you clear all slices belonging to a partition, so those indices must match

Reply:
@sh0rez is correct. You have to always calculate the partition using the stream ID hash and the number of total partitions.

Which unfortunately, will be a bit of a nuisance here in the code.

Reply:
Thinking about it further, by partitioning on streamID, it *does* mean that on average, we'll have up to <numPartitions> "extra" instances of resource metrics, scope metrics, and metrics instances.

IE if you have 8 partitions, and 8 datapoints (with different label sets each). Assuming "ideal" partition hashing, each datapoint will be in a separate partition. Duplicating rm, sm, and m for each partition. This is "fine". But just something to note for the algorithm.


I think we'll have to call `getOrCloneMetric()` inside the loop of `aggregateDataPointsByPartition()`